pkg/agent/plugin/nodeattestor/k8s/k8scert.go

```diff
@@ -0,0 +1,173 @@
+package k8s
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	certificates "k8s.io/api/certificates/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	csrPemBlockType          = "CERTIFICATE REQUEST"
+	ecPrivateKeyPemBlockType = "EC PRIVATE KEY"
+)
+
+func getKubeClient(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath string) (*kubernetes.Clientset, error) {
+	if kubeConfigFilePath == "" {
+		// Try KUBECONFIG env variable
+		kubeConfigFilePath = os.Getenv("KUBECONFIG")
+		if kubeConfigFilePath == "" {
+			// Still no luck, try default (home)
+			home := os.Getenv("HOME")
+			if home != "" {
+				kubeConfigFilePath = path.Join(home, ".kube", "config")
+			}
+		}
+	}
+
+	if kubeConfigFilePath == "" {
+		return nil, fmt.Errorf("Unable to locate kubeconfig")
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error accessing kubeconfig %s: %v", kubeConfigFilePath, err)
+	}
+
+	config.TLSClientConfig.CertFile = clientCertFilePath
+	config.TLSClientConfig.KeyFile = clientKeyFilePath
+	config.TLSClientConfig.CAFile = caCertFilePath
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating clientset: %v", err)
+	}
+	return clientset, nil
+}
+
+func genPrivateKey() (crypto.PrivateKey, error) {
+	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating ECDSA private key: %v", err)
+	}
+	return key, nil
+}
+
+func genCSR(key crypto.PrivateKey, name string) ([]byte, error) {
+	csrTemplate := x509.CertificateRequest{
+		Subject: pkix.Name{
+			// This exact format is required for the CSR to be auto-approved by kube-controller
+			// It must also NOT include any DNSNames or IPAddresses
+			Organization: []string{"system:nodes"},
+			CommonName:   "system:node:" + name,
+		},
+		SignatureAlgorithm: x509.ECDSAWithSHA256,
+	}
+	return x509.CreateCertificateRequest(rand.Reader, &csrTemplate, key)
+}
+
+func getCSRObject(csrBytes []byte, name string) *certificates.CertificateSigningRequest {
+	pemBlock := pem.Block{
+		Type:  csrPemBlockType,
+		Bytes: csrBytes,
+	}
+	pemBytes := pem.EncodeToMemory(&pemBlock)
+
+	ret := &certificates.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Groups:  []string{"system:authenticated"},
+			Request: pemBytes,
+			Usages:  []certificates.KeyUsage{"digital signature", "key encipherment", "client auth"},
```

Comment:
Probably not a big deal since the certificate/key aren't persisted (and thus have a smaller chance of recovery and reuse), but I feel that we should probably only request "digital signature" key usage (if that) to limit the impact if the keypair is somehow recovered. x509pop code was designed to accept arbitrary certificates and so it asserts  `digitalSignature` to make sure that the keypair is ok to be used for generic signature generation as well as whatever normal usage it is intended for. Since the k8s plugin is in complete control of the certificate, that requirement could possible be removed.